R/epi_shift_internal.R

```diff
@@ -0,0 +1,225 @@
+#' Create a shifted predictor
+#'
+#' `step_epi_lag` and `step_epi_ahead` create a *specification* of a recipe step
+#'   that will add new columns of shifted data. The former will created a lag
+#'   column, while the latter will create a lead column. Shifted data will
+#'   by default include NA values where the shift was induced.
+#'   These can be properly removed with [step_epi_naomit()], or you may
+#'   specify an alternative filler value with the `default`
+#'   argument.
+#'
+#' @param recipe A recipe object. The step will be added to the
+#'  sequence of operations for this recipe.
+#' @param ... One or more selector functions to choose variables
+#'  for this step. See [selections()] for more details.
+#' @param role For model terms created by this step, what analysis role should
+#'  they be assigned?
+#' @param trained A logical to indicate if the quantities for
+#'  preprocessing have been estimated.
+#' @param lag,ahead A vector of nonnegative integers. Each specified column will
+#'  be the lag or lead for each value in the vector. The use of negative
+#'  integers will not throw an error and may still work, but is advised against
+#'  as it may have unexpected results. Hence, a warning will be shown if the
+#'  user inputs at least one negative integer value. However, the use of
+#'  non-integer values will throw an error.
+#' @param default Determines what fills empty rows
+#'   left by leading/lagging (defaults to NA).
+#' @param keys A character vector of the keys in an epi_df
+#' @param columns A character string of variable names that will
+#'  be populated (eventually) by the `terms` argument.
+#' @param skip A logical. Should the step be skipped when the
+#'  recipe is baked by [bake()]? While all operations are baked
+#'  when [prep()] is run, some operations may not be able to be
+#'  conducted on new data (e.g. processing the outcome variable(s)).
+#'  Care should be taken when using `skip = TRUE` as it may affect
+#'  the computations for subsequent operations.
+#' @template step-return
+#'
+#' @details The step assumes that the data are already _in the proper sequential
+#'  order_ for shifting.
+#'
+#' The `prefix` and `id` arguments are unchangeable to ensure that the code runs
+#'  properly and to avoid inconsistency with naming. For `step_epi_ahead`, they
+#'  are always set to `"ahead_"` and `"epi_ahead"` respectively, while for
+#'  `step_epi_lag`, they are set to `"lag_"` and `"epi_lag`, respectively.
+#'
+#' @family row operation steps
+#' @rdname step_epi_shift
+#' @export
+#' @examples
+#' epi_recipe(case_death_rate_subset) %>%
+#'   step_epi_ahead(death_rate, ahead = 7) %>%
+#'   step_epi_lag(death_rate, lag = c(0,7,14))
+step_epi_lag <-
+  function(recipe,
+           ...,
+           role = "predictor",
+           trained = FALSE,
+           lag = 1,
+           default = NA,
+           keys = epi_keys(recipe),
+           columns = NULL,
+           skip = FALSE) {
```

Comment:
Despite what I said some weeks ago, we should have `prefix` and `id` here. That's just to be consistent with the typical `recipes::step_*()` functions.

Reply:
I get why this would be the case. However, there would be no infallible way of checking whether something is a lead or lag based on print statements, as a 0 can be a lag or lead here. Should we have that the 0 be treated as a lag, as it is a predictor (given that I am to have errors thrown for negative values)?

Reply:
If the user calls `step_epi_lag()`, then the behaviour is to lag. So the prefix should be `lag_`. Same for `step_epi_lead()`. Then that gets passed directly to `step_epi_shift()`.

Reply:
That is only the default behaviour of the function. However, one could put in `prefix = "asdf_"` or something. If there is no `prefix` argument for lag, then the prefix will be forced as `lag_`.

Reply:
Sorry, maybe there's some misunderstanding. Please put those arguments (with the defaults) into the `step_epi_lag()` and `step_epi_ahead()` functions rather than omitting them there but calling `step_epi_shift()` with those arguments.